src/libasr/pass/template_visitor.cpp

```diff
@@ -0,0 +1,298 @@
+#include <cmath>
+
+#include <libasr/asr_utils.h>
+#include <libasr/pass/pass_utils.h>
+#include <lpython/semantics/semantic_exception.h>
+
+namespace LFortran {
+
+class TemplateVisitor : public PassUtils::PassVisitor<TemplateVisitor>
+{
+public:
+    ASR::asr_t *tmp;
+    std::map<std::string, ASR::ttype_t*> subs;
+    ASR::asr_t *new_function;
+    int new_function_num;
+
+    TemplateVisitor(Allocator &al, std::map<std::string, ASR::ttype_t*> subs,
+            SymbolTable *current_scope, int new_function_num):
+        PassVisitor(al, current_scope),
+        subs{subs},
+        new_function_num{new_function_num}
+        {}
+
+    void visit_TemplateFunction(const ASR::TemplateFunction_t &x) {
+        SymbolTable *parent_scope = current_scope;
+        current_scope = al.make_new<SymbolTable>(parent_scope);
+
+        std::string func_name = x.m_name;
+        func_name = "__lpython_generic_" + func_name + "_" + std::to_string(new_function_num);
```

Comment:
Have you tried `parent_scope->get_unique_name("__lpython_generic_" + func_name + "_")` here?

Reply:
Oh, I did not know about that function before.